draft-ietf-tls-esni.md

```diff
@@ -376,75 +402,63 @@ multiple "outer_extension" extensions with the same extension code point.
 MUST also contain:
 
- - an "esni_nonce" extension
- - a TLS padding {{!RFC7685}}. This SHOULD contain X bytes of padding
-   where X + the actual server name is equal to ESNIConfig.maximum_name_length
-
-Then, the client MUST select one of the server ESNIKeyShareEntry
-values and generate an (EC)DHE share in the matching group. This share
-will then be sent to the server in the "encrypted_client_hello"
-extension and used to derive the SNI encryption key. It does not
-affect the (EC)DHE shared secret used in the TLS key schedule. The
-client MUST also select an appropriate cipher suite from the list of
-suites offered by the server. If the client is unable to select an
-appropriate group or suite it SHOULD ignore that ESNIConfig value and
-MAY attempt to use another value provided by the server. The client
-MUST NOT send encrypted SNI using groups or cipher suites not
-advertised by the server.
-
-When offering an encrypted SNI, the client MUST NOT offer to resume any non-ESNI
+ - an "echo_nonce" extension
+ - TLS padding {{!RFC7685}}. This SHOULD contain X bytes of padding
+   where X + the length of the non-padding bytes in ClientHelloInner
+   is equal to ECHOConfig.maximum_message_length.
+
+Then, the client determines if it supports the server's chosen KEM, as
+identified by ECHOConfig.kem_id. If one is supported, the client MUST
+select an appropriate cipher suite from the list of
+suites offered by the server. If the client does not support the corresponding
+KEM or is unable to select an appropriate group or suite, it SHOULD ignore
+that ECHOConfig value and MAY attempt to use another value provided by
+the server. The client MUST NOT send ECHO using HPKE algorithms not advertised
+by the server.
+
+When offering an ECHO, the client MUST NOT offer to resume any non-ECHO
 PSKs. It additionally MUST NOT offer to resume any sessions for TLS 1.2 or
 below.
 
-Let Z be the DH shared secret derived from a key share in ESNIConfig and the
-corresponding client share in ClientEncryptedSNI.key_share. The SNI encryption key is
-computed from Z as follows:
-
-~~~~
-   Zx = HKDF-Extract(0, Z)
-   key = HKDF-Expand-Label(Zx, KeyLabel, ClientHelloOuter.Random, key_length)
-   iv = HKDF-Expand-Label(Zx, IVLabel, ClientHelloOuter.Random, iv_length)
-~~~~
-
-Where the Hash for HKDF is the hash function associated with the HKDF
-instantiation. The salt argument for HKDF-Extract is a string
-consisting of Hash.length bytes set to zeros. For a client's first
-ClientHello, KeyLabel = "esni key" and IVLabel = "esni iv", whereas
-for a client's second ClientHello, sent in response to a
-HelloRetryRequest, KeyLabel = "hrr esni key" and IVLabel = "hrr esni
-iv". (This label variance is done to prevent nonce re-use since the
-client's ESNI key share, and thus the value of Zx, does not change
-across ClientHello retries.)
+Given an ECHOConfig with fields public_key and kem_id, carrying the the HpkePublicKey and
```

Comment:
s/the the/the/

(Also should public_key and kem_id have quotes? I'm not sure how we typically render all these bits.)

Reply:
I left quotes off for now, reserving them for extension references.